opm/output/eclipse/WindowedArray.hpp

```diff
@@ -0,0 +1,177 @@
+            typename Vector::const_iterator>;
+
+        using Idx = typename Vector::size_type;
+
+        struct NumWindows { Idx value; };
+        struct WindowSize { Idx value; };
+
+        explicit WindowedArray(const NumWindows n, const WindowSize sz)
+            : x_         (n.value * sz.value)
+            , windowSize_(sz.value)
+        {}
+
+        Idx numWindows() const
+        {
+            return this->x_.size() / this->windowSize_;
+        }
+
+        Idx windowSize() const
+        {
+            return this->windowSize_;
+        }
+
+        WriteWindow operator[](const Idx window)
+        {
+            assert ((window < this->numWindows()) &&
+                    "Window ID Out of Bounds");
+
+            auto b = std::begin(this->x_) + window*this->windowSize_;
+            auto e = b + this->windowSize_;
+
+            return { b, e };
+        }
+
+        ReadWindow operator[](const Idx window) const
+        {
+            assert ((window < this->numWindows()) &&
+                    "Window ID Out of Bounds");
+
+            auto b = std::begin(this->x_) + window*this->windowSize_;
+            auto e = b + this->windowSize_;
+
+            return { b, e };
+        }
+
+        const Vector& data() const
+        {
+            return this->x_;
+        }
+
+        Vector getDataDestructively()
+        {
+            return std::move(this->x_);
+        }
+
+    private:
+        std::vector<T> x_;
+
+        Idx windowSize_;
+    };
+
+    template <typename T>
+    class WindowedMatrix
+    {
+    private:
+        using NumWindows  = typename WindowedArray<T>::NumWindows;
+
+    public:
+        using WriteWindow = typename WindowedArray<T>::WriteWindow;
+        using ReadWindow  = typename WindowedArray<T>::ReadWindow;
+        using WindowSize  = typename WindowedArray<T>::WindowSize;
+        using Idx         = typename WindowedArray<T>::Idx;
+
+        struct NumRows { Idx value; };
+        struct NumCols { Idx value; };
+
+        explicit WindowedMatrix(const NumRows& nRows,
+                                const NumCols& nCols,
+                                const WindowSize& sz)
+            : data_   (NumWindows{ nRows.value * nCols.value }, sz)
+            , numCols_(nCols.value)
+        {}
+
+        Idx numCols() const
+        {
+            return this->numCols_;
+        }
+
+        Idx numRows() const
+        {
+            return this->data_.numWindows() / this->numCols();
+        }
+
+        Idx windowSize() const
+        {
+            return this->data_.windowSize();
+        }
+
+        WriteWindow operator()(const Idx row, const Idx col)
+        {
+            return this->data_[ this->i(row, col) ];
+        }
+
+        ReadWindow operator()(const Idx row, const Idx col) const
+        {
+            return this->data_[ this->i(row, col) ];
+        }
+
+        auto data() const
+            -> decltype(std::declval<const WindowedArray<T>>().data())
```

Comment:
While probably correct, I cannot figure out exactly what is the return type here without getting a headache and checking references... I'd rather see a simple `const Vector&` return type here, and a `Vector` for the next function. (Or even `std::vector<T>` instead of `Vector`, which eliminates the need to talk about WindowedArray types in the WindowedMatrix interface -- the compiler would catch related errors in refactoring).

Reply:
> what is the return type here 

"I just return whatever that other function call returns".

One option could be (the somewhat enigmatic) `decltype(auto)`.  Saying `const std::vector<T>&` here is (fairly) benign since reference-to-const implies "returns existing object".  Saying `std::vector<T>` (no qualifiers) for the destructive call is a bit more double-edged.  You'd need to look at the definition of `data_.getDataDestructively()` to become convinced that there's no additional type conversion going on.